qiskit/pulse/schedule.py

```diff
@@ -171,77 +179,135 @@ def _instructions(self, time: int = 0):
                 :class:`~qiskit.pulse.Instruction`
                 starts at and the flattened :class:`~qiskit.pulse.Instruction` s.
         """
-        for insert_time, child_sched in self._children:
+        for insert_time, child_sched in self.children:
             yield from child_sched._instructions(time + insert_time)
 
-    def union(self, *schedules: Union[ScheduleComponent, Tuple[int, ScheduleComponent]],
-              name: Optional[str] = None) -> 'Schedule':
-        """Return a new schedule which is the union of both ``self`` and ``schedules``.
+    def union(self,
+              *schedules: Union[ScheduleComponent, Tuple[int, ScheduleComponent]],
+              name: Optional[str] = None,
+              mutate: bool = False
+              ) -> 'Schedule':
+        """Return a schedule which is the union of both ``self`` and ``schedules``.
 
         Args:
-            *schedules: Schedules to be take the union with this ``Schedule``.
+            schedules: Schedules to be take the union with this ``Schedule``.
             name: Name of the new schedule. Defaults to the name of self.
+            mutate: Perform operation by mutating this schedule. Otherwise return
+                a new ``Schedule``.
         """
         warnings.warn("The union method is deprecated. Use insert with start_time=0.",
                       DeprecationWarning)
+        if mutate:
+            return self._mutable_insert(0, *schedules)
+        else:
+            return self._immutable_insert(0, *schedules, name=name)
+
+    # pylint: disable=arguments-differ
+    def shift(self,
+              time: int,
+              name: Optional[str] = None,
+              mutate: bool = False
```

Comment:
Is it a good idea to confuse users with booleans about mutability?

I know the pulse builder is a better interface for users to access, but I don't think that should mean we can neglect the underlying API completely

Reply:
This or a separate insert is the only way to build up the pulse builder interface to maintain a handle on the schedule being built. We *must* mutate the exposed schedule given the syntax I show.

Reply:
Can you give more details on _must_? The context builder is incapable of updating `pulse_prog` to a newly returned object?

Reply:
Yes, the pulse builder needs to construct a new schedule and yield it when the context is entered. If we do not allow it to be mutated than there is no way to modify this reference we pass out and have the current behavior.

Reply:
One thing of note is that the QuantumCircuit uses `inplace`, in place of `mutate` (see `compose`). I am find with this if we prefer uniformity.

Reply:
Switched to `inplace` in 0dcdc2978.